module/zfs/zfs_ioctl.c

```diff
@@ -1726,6 +1727,131 @@ zfs_ioc_pool_scrub(const char *poolname, nvlist_t *innvl, nvlist_t *outnvl)
 	return (error);
 }
 
+/*
+ * inputs:
+ * dataset		name of dataset
+ * object		object number
+ */
+static const zfs_ioc_key_t zfs_keys_pool_scrub_file[] = {
+	{"dataset",	DATA_TYPE_STRING,	0},
+	{"object",	DATA_TYPE_UINT64,	0},
+};
+
+static void
+scrub_file_traverse(spa_t *spa, blkptr_t *bp, const zbookmark_phys_t *zb)
+{
+	uint64_t blk_birth = bp->blk_birth;
+
+	if (blk_birth == 0)
+		return;
+
+	spa_log_error(spa, zb, &blk_birth);
+	if (BP_GET_LEVEL(bp) > 0 && !BP_IS_HOLE(bp)) {
+		arc_flags_t flags = ARC_FLAG_WAIT;
+		int i;
+		blkptr_t *cbp;
+		int epb = BP_GET_LSIZE(bp) >> SPA_BLKPTRSHIFT;
+		arc_buf_t *buf;
+
+		if (arc_read(NULL, spa, bp, arc_getbuf_func, &buf,
+		    ZIO_PRIORITY_ASYNC_READ, ZIO_FLAG_CANFAIL, &flags, zb))
+			return;
+
+		/* recursively visit blocks below this */
+		cbp = buf->b_data;
+		for (i = 0; i < epb; i++, cbp++) {
+			zbookmark_phys_t czb;
+
+			SET_BOOKMARK(&czb, zb->zb_objset, zb->zb_object,
+			    zb->zb_level - 1,
+			    zb->zb_blkid * epb + i);
+
+			scrub_file_traverse(spa, cbp, &czb);
```

Comment:
I assume this `scrub_file_traverse()` recursion doesn't go deep enough be an issue?

Reply:
If so, we have this problem in other places that use similar methods. So I assume it's not a problem.